Feature/env 134 core component fetch

jeremy-doghouse opened the pull request:
Hey @jorge

One CKAN Client + a few minor fixes.

There is also a fix for chart_export to make it more robust and provide a fallback for Safari

jorgegc-doghouse commented on modules/govcms_ckan_display/govcms_ckan_display.module:
```diff
@@ -70,6 +82,25 @@ function govcms_ckan_display_library() {
     ),
   );
+  $libraries['filesaver'] = array(
+    'title' => 'c3js libraries',
```
Should this be `Filesaver`?

jorgegc-doghouse commented on modules/govcms_ckan_display/govcms_ckan_display.module:
```diff
@@ -70,6 +82,25 @@ function govcms_ckan_display_library() {
   );
+  $libraries['filesaver'] = array(
+    'title' => 'c3js libraries',
+    'website' => 'http://c3js.org',
```
Should this be `https://github.com/eligrey/FileSaver.js/` or something?

jeremy-doghouse replied:
Definately, case of the copy-ala-paste syndrome, will fix

jorgegc-doghouse commented on src/GovCmsCkanClient.inc:
```diff
@@ -0,0 +1,249 @@
+   * @return int
+   *   The response code from the resource, 200 is success.
+   */
+  public function testConnection($endpoint, $resource, $query = array(), $api_key = NULL, $api_version = 3) {
```
Wouldn't it be better if we had the `$api_key` and `$api_version` in the class constructor?

jeremy-doghouse replied:
I was thinking that, but then you would have to pass it each time you instantiated, bit of a pain.
Ideas:
* Can be passed to constructor, but if not, falls back to getting settings
* Settings get obtained in the govcms_ckan_client() helper

jorgegc-doghouse replied:
Gotcha... Why don't we always retrieve the settings in `govcms_ckan_client` and pass them to the class constructor? We will make the assumption that **our module** will always use its settings but still allow other modules/sub-modules to use the client with different settings. How does that sound?

jorgegc-doghouse commented on src/GovCmsCkanClient.inc:
```diff
@@ -0,0 +1,249 @@
+  /**
+   * Populate the settings using drupal variables.
+   */
+  private function getSettings() {
```
This method is making this class a little git rigid mate. It would be nice if we could use this class with different settings other than the module's. Thoughts?

jeremy-doghouse replied:
Best solution for that then is to move settings into the helper `govcms_ckan_client()` then passed in the constructor, you cool with that?

jorgegc-doghouse replied:
YEP! See my previous comment haha!

jeremy-doghouse replied:
Great minds think alike :)

jorgegc-doghouse commented on src/GovCmsCkanClient.inc:
```diff
@@ -0,0 +1,249 @@
+   * Get cached request.
+   *
+   * @return mixed
+   *   If a vaild cache hit found, it is returned, else FALSE on no cache found.
```
Can we please update the docs to reflect what this method is actually doing? The cached value is never returned, only the `$this->responseObject` is being set.

jorgegc-doghouse commented:
Very good PR @jeremy-doghouse thanks. I just got a little bit confused with the `$response` and `$responseObject` int he CKAN class. It would be great if we could unify those as sometimes we use one and something we use the other but given the time constraint I am not even going to suggest we do it of course!

jeremy-doghouse commented:
Hey @jorgegc-doghouse 
Fixed those bits and bobs, rebased off upstream and slipped in a placeholder for our demo page.

re: responseObject, will likely circle back to this but there is some method behind my madness, tried to document in the comments at the top of the file. Main reason was I didn't want to cache everything and wanted to add a few other properties to it (eg. valid, request_time, url params, etc). Anyhoo, can discuss again when have more time.
